src/controlplane/client/authenticator.go

```diff
@@ -0,0 +1,133 @@
+package client
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/newrelic/infra-integrations-sdk/log"
+	"github.com/newrelic/nri-kubernetes/v2/internal/config"
+	v1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/transport"
+)
+
+// Authenticator provides an interface to generate a authorized round tripper.
+type Authenticator interface {
+	Authenticate(endpoint config.Endpoint) (http.RoundTripper, error)
+}
+
+type authenticator struct {
+	logger                  log.Logger
+	secretListerByNamespace map[string]v1.SecretNamespaceLister
+	inClusterConfig         *rest.Config
+}
+
+// NewAuthenticator returns an Authenticator that supports plain, bearer token and mTLS.
+func NewAuthenticator(logger log.Logger, secretListerByNamespace map[string]v1.SecretNamespaceLister, inClusterConfig *rest.Config) Authenticator {
```

Comment:
Logger in the constructor list makes it look a bit boilerplate-y. Could we use an Options pattern, here?

Reply:
Not sure if we have a reason to hide the K8s implementation for Authenticator. Would it be a problem to return it (exported) rather than the interface? Taking advantage of Go's fantastic implicit interfaces :)

Reply:
changed this and the connector to unify the code